### PR TITLE
Swaps with PTLC as beta ledger

### PR DESCRIPTION
**Pull request by luckysori**
It's a little rough, but it works*. The alpha ledger is missing, but it's trivial since we're just gonna manually fund and redeem.

*On my machine all e2e tests started failing with stack overflows. I haven't been able to solve the problem outright, but I managed to mitigate it by running the tests with a `--release` flag and using a specific `runtime` to run the futures (as opposed to annotating tests with `#[tokio::test]`.

Despite these strange errors I don't think there's anything inherently wrong with the implementation. It could be `miniscript-rs` or `enum-as-inner`, but I need help investigating the issue. 

~I suspect the tests will fail on CI~ The tests are failing on CI for the same reason. Interestingly, the tests were passing with a `--release` flag on commit 2dd1d6e on my machine. The more code I add the more likely it is that any test will start failing again with a stack overflow :man_shrugging:

### CODE REVIEW

**Comment by D4nte** [comment]
Don't forget to delete your branches so it can eventually get merged in master.

File: thor/tests/e2e.rs [hunk 1]
```diff
@@ -410,3 +440,213 @@ async fn e2e_force_close_after_updates() {
...
+
+    // Additionally, Alice pays an extra `thor::TX_FEE` to be able to redeem the
+    // PTLC output.
+    let fee_deduction_for_ptlc_redeem = Amount::from_sat(thor::TX_FEE);
+
+    assert_eq!(
+        after_close_balance_alice,
+        after_open_balance_alice + fund_amount_alice + ptlc_amount
+            - fee_deduction_per_balance_output - fee_deduction_for_ptlc_redeem,
+        "Balance after closing channel should equal balance after opening plus PTLC amount, minus transaction fees"
```
Comment (D4nte) [inline thor/tests/e2e.rs:645]:
```suggestion
        "Balance after closing channel should equal balance after opening plus PTLC amount, plus fund amount, minus transaction fees"
```
I found this assertion unintuitive. I think it is safer (in the regards of not loosing money) to that the diff of  balance **before** opening the channel and **after** closing the channel matches the payments done over the channel.

For example. if twice the `fund_amount` is taken from the balance to open the channel, then this assert would not catch the bug and ultimately the user would lose money.

**Review by D4nte — APPROVED** [review]
Well done @luckysori, very impressive 👏

File: thor/tests/e2e.rs (hunk 1 above)
Comment (luckysori) [reply thor/tests/e2e.rs:645]: I agree. I thought it may be difficult to achieve what you describe because we don't control the funding transaction fee, but maybe I was wrong. We should fix it in a bunch of tests, so maybe we could leave it as a TODO and fix it in a follow-up PR.

File: thor/src/lib.rs [hunk 2]
```diff
@@ -395,6 +599,32 @@ impl Channel {
     }
 }
 
+#[allow(clippy::large_enum_variant)]
+#[cfg_attr(feature = "serde", derive(::serde::Serialize, ::serde::Deserialize))]
+#[derive(Clone, Debug, EnumAsInner)]
+pub enum ChannelStateKind {
```
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/lib.rs:605]: I did struggle to name/model this. By `ChannelStateKind::BalanceOutputs` I meant a channel state that only contains balance outputs in the split transaction.

File: thor/src/lib.rs [hunk 3]
```diff
@@ -178,66 +184,217 @@ impl Channel {
...
+                vec![balance_output_self, balance_output_other],
+                TX_s_time_lock,
+            )
+            .await?;
+
+            Ok(())
+        })
     }
 
     pub async fn _update<T>(
```
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/lib.rs:400]: Definitely not!

File: thor/src/lib.rs [hunk 4]
```diff
@@ -178,66 +184,217 @@ impl Channel {
 
     /// Perform an atomic swap with a thor channel as beta ledger in the
     /// role of Alice.
-    pub async fn add_ptlc_alice<T>(
+    #[allow(clippy::too_many_arguments)]
+    pub async fn add_ptlc_alice<T, W>(
```
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/lib.rs:188]: Very true. I am not sure if there should be a general purpose `add-ptlc` API anyway. I'm not sure how the function body compares to adding PTLCs for multi-hops.

File: thor/src/lib.rs [hunk 5]
```diff
@@ -178,66 +184,217 @@ impl Channel {
...
-        // function).
-        // 4. Receive Bob's signature for ptlc_refund and encsignature for ptlc_redeem.
-        // 5. Run channel update protocol to add PTLC output.
-        // 6. Send encryption_secret_key to Bob (who will use it to redeem alpha asset).
-        // 7. Attempt to perform a channel update to add PTLC output to Alice's balance
-        // output. If Bob doesn't respond, force close and publish ptlc_redeem. If he
-        // does respond, carry out the update and feel free to forget the
+        // 3. Receive Bob's signature for ptlc_refund and encsignature for ptlc_redeem.
+        // 4. Run channel update protocol to add PTLC output.
+        self._update(
```
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/lib.rs:235]: I think you can also use them to differentiate between the public and private "version" of a function, no?

File: thor/src/lib.rs [hunk 6]
```diff
@@ -55,7 +61,7 @@ pub struct Channel {
     final_address_self: Address,
     final_address_other: Address,
     TX_f_body: FundingTransaction,
-    current_state: ChannelState,
+    current_state: ChannelStateKind,
```
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/lib.rs:64]: Yeah, I'm not sold on kind here either.

File: thor/src/lib.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply thor/src/lib.rs:235]: It's more of a C++/Java convention than it is Rust.

File: thor/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (D4nte) [reply thor/src/lib.rs:605]: I see, maybe `standard` and `WithPTLC` as PTLC is kind of the exception here.
or `BalanceOutputsOnly`

"Balance output" is not great. What we mean is that in the split transaction, there is one output for each party, locked against a pkh, right?

But with ptlc, then we have a 3rd output? Can you describe a bit more and I can see if I can help.

File: thor/src/keys.rs [hunk 7]
```diff
@@ -260,7 +271,7 @@ fn random_key_pair() -> (Scalar, Point) {
     (secret_key, public_key)
 }
 
-fn public_key(secret_key: &Scalar) -> Point {
+pub(crate) fn public_key(secret_key: &Scalar) -> Point {
```
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/keys.rs:274]: This is temporary. I would instead introduce a `PtlcSecret` and `PtlcPoint` type. I left a TODO somewhere else, but I should have left a comment here.

File: thor/tests/e2e.rs (hunk 1 above)
Comment (D4nte) [reply thor/tests/e2e.rs:645]: Have a look at my PR: #49, I did that with @bonomat's help.

File: thor/src/keys.rs [hunk 8]
```diff
@@ -39,9 +39,14 @@ impl OwnershipKeyPair {
     }
 
     pub fn encsign(&self, Y: PublishingPublicKey, digest: SigHash) -> EncryptedSignature {
+        self.encsign_with_point(Y.0, digest)
+    }
+
+    // TODO: Introduce a type to model secret and point?
+    pub fn encsign_with_point(&self, Y: Point, digest: SigHash) -> EncryptedSignature {
```
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/keys.rs:46]: It's just for the PTLC point. We just need a type for it and this will go away.

File: thor/tests/e2e.rs (hunk 1 above)
Comment (luckysori) [reply thor/tests/e2e.rs:636]: Not gonna do it now, but we should definitely align all tests to do the same thing at some point.

File: thor/src/lib.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (luckysori) [reply thor/src/lib.rs]: I'm happy with `ChannelState::{Standard, WithPtlc}` and the inner one being `StandardChannelState` for now. Feel free to rework at any time.
